src/directive/rateLimit.ts

```diff
@@ -9,58 +12,132 @@ import {
 import { GraphQLError } from 'graphql';
 import { singleRedisClient } from '../redis';
 import { Context } from '../Context';
+import { logger } from '../logger';
+import { WATERCOOLER_ID } from '../common';
 
 export class CustomRateLimiterRedis extends RateLimiterRedis {
-  constructor(props) {
+  constructor(props: IRateLimiterRedisOptions) {
     super(props);
   }
 
   // Currently not doing any special actions/overrides
   // This was primarily introduced to make debugging easier by logging the details of rate limited queries/mutations after receiving a request
-  consume(key, pointsToConsume, options) {
+  consume(
+    key: string | number,
+    pointsToConsume?: number,
+    options?: { [key: string]: unknown },
+  ) {
     if (process.env.NODE_ENV === 'development') {
-      console.log(`[CONSUME] ${key} for ${pointsToConsume}`);
+      logger.info(`[CONSUME] ${key} for ${pointsToConsume}`);
     }
 
     return super.consume(key, pointsToConsume, options);
   }
 }
 
-const keyGenerator = (directiveArgs, source, args, context, info) =>
-  `${context.userId ?? context.trackingId}:${defaultKeyGenerator(
-    directiveArgs,
-    source,
-    args,
-    context,
-    info,
-  )}`;
+const keyGenerator: RateLimitKeyGenerator<Context> = (
+  directiveArgs,
+  source,
+  args,
+  context,
+  info,
+) => {
+  {
+    switch (info.fieldName) {
+      case 'createFreeformPost':
+      case 'submitExternalLink':
+      case 'sharePost':
+        return `${context.userId ?? context.trackingId}:createPost`;
+      case 'commentOnPost':
+      case 'commentOnComment':
+        `${context.userId ?? context.trackingId}:createComment`;
+      default:
+        return `${context.userId ?? context.trackingId}:${defaultKeyGenerator(
+          directiveArgs,
+          source,
+          args,
+          context,
+          info,
+        )}`;
+    }
```

Comment:
The default key looks something like this `rateLimit:userId:Mutation: createFreeformPost`, so it would then limit you to 10 free form posts per hour, but you could still share links when you exceed freeform limit.

Reply:
Ah I see, you are basically overwriting so that different mutations return the same key.

Reply:
I could split the key generator into different ones, and then create a new rate limit directive for posts and comments, i.e. so it would be `@commentRateLimit(limit: 10, duration: 3600)`, which would use the key generator for comments. But it would become a lot of code, that's why I went with a switch statement here and in the throw error.

Ideally, this could've been set in the directive itself, but would mean we fork it and adjust it ourselves.

Reply:
For now this does not sound like something worth investing time into and then also maintaining the fork. But we can maybe contribute to the repo directly to add support for what we need FYI @rebelchris

Reply:
That was my conclusion as well, hence why I tried to find a way to do it without modifying the package

Reply:
Yeah wouldn't put too much effort in it.
Let's start with this and see how many times the limits even hit?